### ISSUE TITLE
Allow custom (nonstandard) attributes.

### ISSUE THREAD

**Issue by steida** [facebook/react#140]
Various frameworks uses custom attributes. React could allow to extend default data- and aria- prefixes. Something like this:

``` javascript
React.registerCustomAttributePrefix('e-');
React.registerCustomAttributePrefix('ng-');
```

**Comment by tehnomaag** [comment]
I ran across this "custom" attributes issue too, but with trying to render (standard) SVG. So I thought I'll add my real world example to this issue.

I was trying to use masks like so:

```
render: function () {
return <svg>
  <defs>
    <mask id="my-mask">
      <rect .../>
    </mask>
  </defs>
  <g mask="url(#my-mask)">
    ...
  </g>
</svg>;
}
```

First problem is that to make the `url(#...)` syntax work you have to have `xmlns:xlink="http://www.w3.org/1999/xlink"` attribute set on `<svg>` tag. And that is not whitelisted attr. OK I can add that with `setAttribute()` call in a `componentDidMount` function if needed.

But also the `mask` attribute is not whitelisted by React, so I'd need to go and modify all those DOM nodes one by one as well if I had many masked items rendered there.

Hopefully this shows that there are other and "standard" attributes that React is missing support for.

**Comment by dandelany** [comment]
+1 on some sort of solution to this - I think I prefer no attributes whitelist at all. There are a huge number of possible SVG attributes, many of which are currently unsupported. This makes it quite a pain to do anything beyond the most basic SVG rendering with React - which is a shame because I think SVG should be in React's wheelhouse. I've definitely run into the "mask" issue mentioned above before; this time though it's "markerWidth" and "markerHeight" that are causing problems for me.

**Comment by jstrimpel** [comment]
Would it be possible to introduce the concept of a [config](https://gist.github.com/jstrimpel/93c23f5ce249f8a48da9) to support customization like this? Seems like this could be expanded to solve the SVG attributes issues as well by making `isCustomAttribute` for the SVG attributes config match all strings.

**Comment by jsfb** [comment]
@jstrimpel We want to avoid adding "configs" because they make things more complicated for users (more api surface area, more things that could go wrong / introduce component incompatibilities) and they make it more difficult for us (more permutations of conditions need to be tested).  We know we want to get rid of the attributes whitelist anyway, so it makes no sense to introduce a "config" and then take it away in the subsequent release.

**Comment by jsfb** [comment]
Oops, didn't mean to close.

**Comment by jstrimpel** [comment]
@jsfb Thanks. +1 for removing the whitelist. Is there an ETA on that?

**Comment by geelen** [comment]
Awesome, removing the whitelist will fix my use case perfectly. Very interested to see what SVG wizardry you're cooking up @tehnomaag & @dandelany, I hadn't considered using React for SVG but it could be pretty cool!

**Comment by jhicken** [comment]
+1 For removing the whitelist.

**Comment by Aaronius** [comment]
I appreciate the removal of the tag name whitelist. That has allowed us to support many custom elements however it has not allowed us to support custom elements that extend from native elements which require the `is` attribute as described in my [previous comment](https://github.com/facebook/react/issues/140#issuecomment-65161467). A resolution on this would be super-duper-appreciated.

**Comment by spicyj** [comment]
@Aaronius It's something we want to do. Until then, I'm happy to take a PR adding "is".

**Comment by jhicken** [comment]
@spicyj I would totally do a PR for something small like adding `is` but I don't think it would solve the entire custom element problem. So its more than just the `is` attribute. For example if you use the `is` attribute to extend an element and your extension uses a custom attribute named `numberofplops` then `numberofplops` would be removed and your custom element may fail to behave properly.

I can get around it for now using @nfroidure 's method but it make me really sad to do that. So any info about priority to remove the whitelist would be rad.

**Comment by jsfb** [comment]
@jhicken Currently we skip the whitelisting check if the component name contains a dash.  To enable component inheritance, you would need to also skip the whitelist check if the component contained an `is` attribute.

**Comment by spicyj** [comment]
@jsfb Do we? I know we talked about it but that's not what the code looks like:

https://github.com/facebook/react/blob/0183f70797183ae5371f61a40c1c13991cd7b104/src/browser/ui/ReactDOMComponent.js#L435-L441

**Comment by jhicken** [comment]
I just threw a commit together that does what @jsfb suggested. Have a peak let me know your thoughts.

**Comment by Gozala** [comment]
This would resolve an issue I've being struggling with for a while (see #2746) for details

**Comment by jhicken** [comment]
So the now this stuff works on master I found an issue where if your using a custom component react does not translate className to class.

EX:
`<awesome-sauce className="fluffy" />`

becomes

`<awesome-sauce classname="fluffy" />`

instead of 

`<awesome-sauce class="fluffy" />`

What file makes this translation?

**Comment by ljharb** [comment]
This issue has been open for about 2 years - is there any chance of being able to use any attributes I wish in jsx for a general case?

**Comment by sebmarkbage** [comment]
@ljharb For custom elements, I.e. tags with dashes in them, yes. That has landed in master. For other tags, no. What is the use case? Is it an attribute with a dash in it? We might be able to support that.

@jhicken Custom elements uses "attributes" where as HTML nodes uses "properties". That's why they have different names. We don't intend to change that. It is just a consequence of custom elements relying heavily on attributes but the proper way to interact with DOM is through properties.

**Comment by ljharb** [comment]
No, no dashes. See https://help.pinterest.com/en/articles/prevent-pinning-your-site - the "nopin" attribute is what Pinterest requires to exempt an image from showing up in the "pin" dialog. This is nonstandard (shame on them), but has no dashes, and isn't something I can control.

How can I use this in jsx without creating a custom element?

**Comment by jhicken** [comment]
@ljharb You could use the the ```is``` attribute to force react to ignore the whitelist. The ```is``` attribute ignoring is also new in master. I think thats a really hacky way to use the ```is``` attribute but It would work. Just remember that if you put a class on the element with the ```is``` attribute you have to use ```class``` not ```className```.

Like this.
```<img src="foo.jpg" is nopin="nopin" class="dontTazeMeBro" />```

@sebmarkbage Thats too bad... I don't know enough about the separation of how react deals with attributes vs properties under the covers to really present a real solution to this problem. However I do see this as an api issue. I think as custom components become more popular. This is going to become really annoying to have a goofy mix of class and className attributes in your jsx. (I say this because I'm working on a project that makes heavy use of custom components) So this may be a naive question but, would it be possible when ```-``` is in element name or ```is``` attribute exist.. we convert the elements html attributes to react properties?

**Comment by sebmarkbage** [comment]
@jhicken We don't know what `className` might mean for a custom component. That might be the name of a legit attribute so if we were to convert it, we would now do some weird magic work that breaks that API. Unlike properties, attributes can only be strings. Additionally, you'll notice that event listeners won't work for these.

It is the opinion of the designers of web components and the spec that attributes is a mistake and should not be used directly. It is a serialized form. It shouldn't be used directly for anything other than serialized HTML, and then properties should take over. However, this is not really how most custom elements are consumed. The public APIs right now are mostly attributes. If properties are made available they're typically imperative methods that doesn't translate well to React's declarative API. We're forced to use attributes since it is the only thing that is possible for custom elements.

For consistency, the only solution would be to revert what we do for normal HTML elements and have them apply attributes instead of properties. That is more intuitive but strictly inferior since it doesn't allow us to properly model things like `classList`, `transform`, `defaultValue` vs. `value`, `checked` etc. These are not just attribute strings, they come with richer data structures or richer behavior differences than the serialized form allows for.

More over, it is our opinion (and the opinion of the React, Ember and Angular teams) that custom elements (Web Components) is a flawed model for composition and it won't be treated as first-class in either framework. There are too many compromises to make that work flawlessly.

We don't want to compromise the primary HTML API to support the custom elements edge case. So unfortunately, that's how we end up with this subpar API.

The recommended solution is to wrap your Web Components in a React component that can provide a richer API and use refs to interact with events / properties / methods. As part of that you can always translate the names as you see fit `<my-component class={this.props.className} />`.

@ljharb We would like to support a non-whitelist solution but not quite sure how to tackle it safely right now. As a workaround you can always get a `ref` on an element and call `setAttribute('nopin', '')` on it yourself. Ofc, this is annoying to do many times but best practice is to build small reusable components that you compose, if you do that then you should only need to use this hack once.

Btw, I will address these issues in my React Europe talk on July 2nd.

**Comment by dantman** [comment]
@sebmarkbage 
> @ljharb We would like to support a non-whitelist solution but not quite sure how to tackle it safely right now. As a workaround you can always get a ref on an element and call setAttribute('nopin', '') on it yourself. Ofc, this is annoying to do many times but best practice is to build small reusable components that you compose, if you do that then you should only need to use this hack once.

This probably won't work in this case, as Pintrest is probably not executing client side JS and instead simply statically analyzing `<img>` tags for a `nopin` attribute.

**Comment by sebmarkbage** [comment]
Oh, this is for server-side rendering? Yea, that sucks. Maybe Pinterest should be nicer community members and use the `data-` namespace like it was intended. :(

**Comment by spicyj** [comment]
Oh, I assumed this was for the browser extension that adds "Pin It" buttons to everything.

**Comment by dantman** [comment]
> Oh, this is for server-side rendering? Yea, that sucks. Maybe Pinterest should be nicer community members and use the data- namespace like it was intended. :( 

Yeah, for that use case. Though data attributes also sound wrong. This would be easier if `<img>` had something like `rel` so `rel="nopin"` would be viable.

**Comment by spicyj** [comment]
`data-pin-nopin="true"`: https://github.com/pinterest/widgets/issues/29

**Comment by sebmarkbage** [comment]
Nice @spicyj! Problem solved. I figured I should've checked for alternatives but I'm sick in bed and also lazy.

**Comment by Gozala** [comment]
@sebmarkbage @spicyj this maybe solving this specific instance but it is still a very painful issue for some of us. We are basically forced to do a horrible hacks to workaround lack of general attribute support. I have wrote about this in #2746 and also explained workardounds we employ.

I would very much like for this issue to be resloved maybe for the very least react can be made configurable so users could extend the whitelist.

**Comment by spicyj** [comment]
@Gozala Yes, we understand. That's why this issue is still open.

**Comment by pencilcheck** [comment]
Using angular material with react and attributes such as "layout" that are introduced in angular material doesn't work. How do I whitelisted those attributes for the moment? https://material.angularjs.org/latest/#/layout/container

**Comment by chromakode** [comment]
Based on @nfroidure's approach above, for server-side rendering (e.g. rendering archaic table attributes in HTML emails), here's a workaround I'm experimenting with using in the meantime:

```js
var DOMProperty = require('react/lib/ReactInjection').DOMProperty
DOMProperty.injectDOMPropertyConfig({
  Properties: {
    'valign': DOMProperty.MUST_USE_ATTRIBUTE,
  }
})
```

**Comment by terinjokes** [comment]
With 0.14 splitting `react-dom` from `react`, could an alternative module be created to allow these archaic HTML attributes (`react-dom-emails` or something of the like)?

**Comment by IL55** [comment]
I just want to notify that in React  0.13.3, still no svg support for "mask" attribute. Please don't mixed up with mask tag, i.e. next is allowed:
```
      <defs>
           <mask id="myMask"> .... </mask>
      </defs>
```
but using mask is not possible (react 0.13.3 removes that attribute)
```
    <g mask="url(#myMask)"> ... </g>
```

Here is jsfiddle example:
https://jsfiddle.net/kbcpLyax/3/

More details could be found in next comment:
https://github.com/facebook/react/issues/140#issuecomment-73788054

**Comment by yaycmyk** [comment]
@IL55 i'd open a separate ticket for it

**Comment by Deepblue129** [comment]
React 13.3 strips my Animate tag!

From: 
```html
<animate attributeName="stop-color" dur="24s" values="rgba(0,0,12,0);rgba(25,22,33,.3);rgba(32,32,44,.8);rgb(58,58,82);rgb(81,81,117);rgb(138,118,171);rgb(205,130,160);rgb(234,176,209);rgb(235,178,177);rgb(177,181,234);rgb(148,223,255);rgb(103,209,251);rgb(56,163,209);rgb(36,111,168);rgb(30,82,142);rgb(91,121,131);rgb(157,166,113);rgb(233,206,93);rgb(178,99,57);rgb(47,17,7);rgb(36,14,3);rgb(47,17,7);rgba(75,29,6,.4);rgba(21,8,0,0);rgba(0,0,12,0)" repeatCount="indefinite"></animate>
```

To:
```html
<animate data-reactid=".0.0.1.$=10:0.$=10:0.$=10:0.$=11:0"></animate>
```

**Comment by suryagaddipati** [comment]
@Deepblue129 you can do `ref.getDomNode().setAttribute(...`  in component lifecycle method.

**Comment by despairblue** [comment]
This also makes it impossible to use react to generate x3d scenes like

```jsx
<transform>
  <inline url={url} namespace={url}/>
</transform>
```

and having these rendered using [x3dom](www.x3dom.org) or [later natively by the browser](https://www.w3.org/community/declarative3d/)

Most attributes will be stripped from the rendered content, like `translations` or `rotations`, or even `url` for the inline node above.

**Comment by almirfilho** [comment]
It doesn't allow `tabindex` attribute too. So it's impossible to follow the W3C [WCAG Spec SCR29](http://www.w3.org/WAI/GL/WCAG20/WD-WCAG20-TECHS/SCR29.html) to enhance accessibility in non focusable elements by default, like a simple `<div>`.

**Comment by syranide** [comment]
@almirfilho `tabIndex` and it'll work just fine.

**Comment by almirfilho** [comment]
@syranide thanks for the tip!

**Comment by eyworldwide** [comment]
@chromakode thx! it works in my project...

**Comment by chrissheppard41** [comment]
Hi,

Might be worth taking a look at supporting googles Rich snippet https://developers.google.com/structured-data/rich-snippets/?hl=en

Just a possible idea

**Comment by dantman** [comment]
@chrissheppard41 
> Might be worth taking a look at supporting googles Rich snippet https://developers.google.com/structured-data/rich-snippets/?hl=en

Microdata, RDFa, and script tags are already supported.

**Comment by chrissheppard41** [comment]
@dantman 

Yeah I read this: https://facebook.github.io/react/docs/tags-and-attributes.html

But when I embed say "itemscope" or any others like so:

```
    render() {
        return (
            <div ref="parent" className="review_feed" itemscope="" itemtype="http://schema.org/Product">
                ...Something here
            </div>
        );
    }
```

It prints out:

```
<div class="review_feed" data-reactid=".0">
    ...Something here
</div>
```

It's not applying to the dom

This is likely to my misunderstanding but some clarity would be lovely. Thanks.

**Comment by dantman** [comment]
@chrissheppard41 That's because React-JSX is not HTML. You converted the html `class` to React's `className` but didn't do the same for the rest of the attributes.

You need to [convert](https://facebook.github.io/react/html-jsx.html) the HTML examples to React-JSX:
```html
<div class="review_feed" itemscope="" itemtype="http://schema.org/Product">
  ...Something here
</div>
```
Becomes:
```jsx
<div itemType="http://schema.org/Product" itemScope className="review_feed">
  ...Something here
</div>
```

**Comment by chrissheppard41** [comment]
Ah, I'm sorry, that worked, Thank you.

**Comment by monolithed** [comment]
I really don't understand why some library should restrict my application architecture?

**Comment by antitoxic** [comment]
@spicyj @sebmarkbage @syranide

So far, from what I've read, this looks like a real problem, not just a feature perk.

Is there a downside for React, *before an ideal solution is provided*, to supply some  intermediate one?

Like:

```html
<div thirdpary-attr='"i need this" allowAttrs="thirdpary-attr">
<!-- or previously defined -->
<div thirdpary-attr='"i need this" allowAttrs={this.allowedAttrs}/>
<!-- or -->
<div thirdpary-attr='"i need this" safeAttrs="thirdpary-attr">
```

I understand that the above is not the wisest thing to do with an API (introduce a temporary interface) bu the issue is 2 years old and even the best APIs I know change in that time.

If however if you completely disagree, could React documentation include a workaround? That will be super-useful.

**Comment by antitoxic** [comment]
PS: Aside from use-cases given above (*+ things others like https://github.com/facebook/react/issues/3809*) I'm using attributes for something that people can call me heretic - simply for clear way of styling (*i.e.* `.myCSSComponentClass[size="big"]`) . Combining attributes and their values has proven to be way more readable than just throwing classes in.

For that purpose I'm using a simple ES7 decorator.

`decorators.js`: 
```js
import ReactDOM from 'react-dom';

export function customAttributes(...attrs) {
    return function(component) {
        var originalDidMount = component.prototype.componentDidMount;
        component.prototype.componentDidMount = function() {
            if (originalDidMount) originalDidMount .apply(this, arguments);
            var rootDom = React.findDOMNode(this);
            attrs.map(attr => {
                if (!this.props.hasOwnProperty(attr)) return;
                rootDom.setAttribute(attr, this.props[attr])
            });
        };
        return component;
    }
}
```

```js
import React from 'react';
import {customAttributes} from 'decorators';

@customAttributes('color', 'size')
export class Button extends React.Component {
...
}
```

To me ot looks like a clean approach. Am I doing something terribly akward? Is this a reccomended approach?

**Comment by ajfarkas** [comment]
Looking forward to this fix. Would like to use SVG Icons, but dangerously setting `xlink:href` is a bummer.

**Comment by spicyj** [comment]
@ajfarkas `xlinkHref={...}` works already in 0.14.

**Comment by ajfarkas** [comment]
true, but `<use>` tags don't.

I appreciate the attempt at giving people a heads up, but this is producing a major headache, since I have to change all of my SVG code, then will have to change it back once this gets updated.

**Comment by spicyj** [comment]
`<use>` tags (and all SVG tags and attributes) should work in master and will work in v15.

What do you mean about changing your code?

**Comment by tnrich** [comment]
Hey there,
Props for making react, its awesome! Pun intended :)

I'm trying to use the "inert" html attribute to disable a div (and polyfilling it so it works on all browsers), but react doesn't seem to be letting it through. Is there any way to get this to work currently?

Thanks!

**Comment by gaearon** [comment]
@tnrich

For not yet supported attributes, you can put `ref` on the element and call `setAttribute()` manually.

For example:

```js
return <div ref={node => node && node.setAttribute('inert', '')} />
```

**Comment by tnrich** [comment]
Awesome thanks @gaeron !

**Comment by dantman** [comment]
@tnrich `inert` was [dropped from the standard drafts](https://github.com/whatwg/html/commit/5ddfc78b1f82e86cc202d72ccc752a0e15f1e4ad) some time ago; it's not a pending standard.

**Comment by tnrich** [comment]
@dantman, that may be so, but, is there a reason not to use it with a polyfill (kind of a misnomer in this case), given the listed caveats? 

Here's the npm module I'm using: https://github.com/GoogleChrome/inert-polyfill

Sorry to stray off topic everyone..

**Comment by fritx** [comment]
> If you want to use a custom attribute, you should prefix it with data-.
- https://facebook.github.io/react/docs/jsx-gotchas.html#custom-html-attributes

But seems nwjs only recognizes `nwsaveas`, `nwworkingdir`, etc.
- https://github.com/nwjs/nw.js/wiki/File-dialogs#save-file

What could I do with this?
Thanks!

**Comment by jayphelps** [comment]
@fritx https://github.com/facebook/react/issues/140#issuecomment-62527758

**Comment by fritx** [comment]
@jayphelps thanks for pointing out the comment (there were too many to figure out)

> until there's movement on this.

So there is still some movement on this?

I had tried the workaround and it did work util later. But I've found another issue just now https://github.com/nwjs/nw.js/issues/3372#issuecomment-147936759

Thanks to that buddy's great effort. I guess it's some problem of nwjs itself.

**Comment by fritx** [comment]
I still wonder is `a.setAttribute('xxx')` entirely equivalent to `<a xxx>` stuff, or there is still some tiny difference between the two definition?

Btw, [git-blame-someone-else](https://github.com/jayphelps/git-blame-someone-else) is awesome. I just plan to fake some funny commit in my own prj :smile: @janhancic

**Comment by jayphelps** [comment]
@fritx using `setAttribute` inside `componentDidMount` is different in the sense that it happens _after_ the element has already been inserted into the DOM. Depending on how that library works, that might impact things.

Often, when you decide it's worth it to integrate a library that is incompatible with React, it's best to just create a wrapper React Component and then just create the real nodes/use plugins and insert them inside `componentDidMount`, just like you'd do with regular JavaScript/jQuery; but remember to clean them up in `componentWillUnmount`. This isn't ideal, but I personally take a "reality isn't always ideal" approach to getting work done.

**Comment by fritx** [comment]
> Btw, git-blame-someone-else is awesome.

**Comment by fritx** [comment]
> Btw, git-blame-someone-else is awesome.

As a bash beginner, I had to do kind of [research](https://github.com/fritx/git-green-someone-else#related), util I eventually created [git-green-someone-else](https://github.com/fritx/git-green-someone-else), which was heavily inspired by git-blame-someone-else.

Hope any guy will enjoy it!
And if I'm missing something, let me know.

PS: I'm fond of taking friends --force to my project so it looks active LOL..

**Comment by poscar** [comment]
Another use for this is for using React to perform server side rendering of Google AMP pages. Google AMP requires a ```<html ⚡></html>``` or ```<html amp></html>```. More info [here](https://github.com/ampproject/amphtml/blob/master/spec/amp-html-format.md#required-markup). 

Currently react removes the amp attributes.

Thoughts?

**Comment by dantman** [comment]
I think the `ReactInjection.DOMProperty` interface is really nice. In the future if that interface was polished and made into a public interface when the devs are happy with its stability; I think it would make for a very good solution to this issue.

The range of nonstandard properties with reason to be used in react is fairly narrow. If we had that interface, then we could just create libraries that would inject support for nonstandard properties and attributes into react.

A `react-nw` library could inject `nw*` attributes; a `react-amp` library could inject `amp` attributes; and a `react-angular` library could not just support `ng-*` attributes, but perhaps support them as actual properties such as `ngModel`. And these kinds of libraries could do more than just whitelisting attributes, they could better integrate React with the library they're tied to.

**Comment by sebmarkbage** [comment]
What if there's both `react-angular` and `react-ng` libraries both integrating using the same attributes?

**Comment by dantman** [comment]
@sebmarkbage Then they'd be conflicting libraries and you could only use one of them. Just like if you tried to use Angular twice or another library that used `ng-*` attributes in the dom; or if you used two React components that defined and used a string type context parameter named `foo` for two different purposes.

**Comment by poscar** [comment]
Sadly the following snippet does not work for me:
```
const DOMProperty = require('react/lib/ReactInjection').DOMProperty;
DOMProperty.injectDOMPropertyConfig({
  Properties: {
    amp: DOMProperty.MUST_USE_ATTRIBUTE
  },
  isCustomAttribute: (attributeName) => {
    return attributeName === 'amp';
  }
});
```

Any ideas why?

I'll have to resort to doing string replacement, if I can't get React to allow this attribute. :(

**Comment by sebmarkbage** [comment]
The server renderer not having an alternative way to fix this particular problem is unfortunate. We should fix that in isolation. Spawned a new issue. https://github.com/facebook/react/issues/6798

**Comment by benwiley4000** [comment]
Respectfully, why hasn't this been fixed yet? It's a bug. React can reserve its own attributes (like `className` and leave the rest open. If something breaks in a future React update, you can lock your React version or do what everyone else does, read the documentation, and upgrade with a fix.

`data-` is a convention but it is not enforceable. Bootstrap, for instance, introduces custom attributes like [`readonly`](http://v4-alpha.getbootstrap.com/components/forms/#readonly-inputs) and those need to be supportable. This is not an edge case.

**Comment by SunnyGurnani** [comment]
Is there any update on this?

**Comment by benjdlambert** [comment]
@jayphelps is 100% spot on.

I had a real pain trying to work out a shocker of a bug recently.
With the addition of the new `playsinline` to iOS10, if you put it in `componentDidMount` and the `src` attribute is already on the element from the `render()` it will not play inline.

I guess its kind of a bug from Safari point of view, but this is just one use case.

There will be other cases where adding elements after the element has already been inserted into the DOM which cause issues.

I fixed this, by using the `is` property on the element, which removes validation of props. 

/blam

**Comment by jayphelps** [comment]
@benjdlambert I think you meant someone else, my only comment was linking to someone elses comment. 😄

**Comment by benjdlambert** [comment]
@jayphelps I meant the comment after your link to the comment 👍

**Comment by nhunzaker** [comment]
FWIW, I've been maintaining (sort of) a PR to remove the attribute whitelist:

https://github.com/facebook/react/pull/7311

I don't speak for the React team but, hopefully, if the decision is made to support custom attributes by eliminating the whitelist, this handles much of the legwork.

**Comment by antitoxic** [comment]
In continuation of what @gaearon [said](#issuecomment-186010590) in Feb this year, if anyone wants to make custom attributes a bit more tidy to write, you can do it with a babel plugin (*or other AST-based transformer if you're not using babel*). 

Here's a quick working example (*you can change the `attrs` to what you find appropriate*):

```js
// your view
function MyComponent({size, density, href}) {
    return <a href={href} attrs={{size, density}}>Go</a>
}
```

```js
// the babel plugin
var template = require('babel-template');

var refBody = template(`
    var attrs = ATTRS_MAP;
    Object.keys(attrs).forEach(function(k) {
      n.setAttribute(k, attrs[k])
    })
`);

module.exports = function ({types: t}) {
  var visitor = {
    JSXAttribute(path) {
      if (path.node.name.name != 'attrs') return;
      path.replaceWith(
        t.JSXAttribute(
          t.JSXIdentifier('ref'),
          t.JSXExpressionContainer(
            t.functionExpression(
              null,
              [t.Identifier('n')],
              t.blockStatement(refBody({
                ATTRS_MAP: path.node.value.expression,
              }))
            )
          )
        )
      );
    }
  };

  return {
    inherits: require("babel-plugin-syntax-jsx"),
    visitor: visitor
  };
};
```

Which will generate the required `ref` logic and output:
```js
function MyComponent(_ref) {
	    var size = _ref.size;
	    var density = _ref.density;
	    var href= _ref.href;

	    return React.createElement(
	        'a',
	        { href: href, ref: function ref(n) {
	                var attrs = { size: size, density: density };
	                Object.keys(attrs).forEach(function (k) {
	                    n.setAttribute(k, attrs[k]);
	                });
	            }
	        },
	        'Go'
	    );
	};
```

**Comment by Gozala** [comment]
Problem with a `ref` based solution is that `ref` is called after element is already in the document tree, unfortunately sometimes you need to set attributes before element is in the document tree or behavior is going to be different.

**Comment by vizath** [comment]
Note that since 15.4, `ReactInjection` is not part of React anymore, but of ReactDOM

```js
const DOMProperty = require('react-dom/lib/ReactInjection').DOMProperty;
```

**Comment by joh1234** [comment]
Update README
https://goo.gl/JBljdR

**Comment by alivedise** [comment]
It's about 2017 and this is still opened. I understand that you don't think property and attribute is the same thing, but why not just treat all custom attribute on native element as legal and treat all unknown property on react element as 'property'?

If you can't catch up all the needs of new attributes on time, I'm afraid you have to open it to the developers. It makes me feel React is telling us using custom attribute is guilty. But why? Is Web standards telling us don't use that? It's indeed a real requirement.

**Comment by AntonioRedondo** [comment]
There are many valid use cases above described where it should be possible to add custom attributes. I've came accross with another one: to attach `onReadyStateChange` to `<iframes>` so that I can track when the `<iframe>` starts loading the embedded page. You cannot do this with `element.addAttribute()` since the element isn't accesible when it starts loading after the `src` attribute is parsed).

**Comment by sebmarkbage** [comment]
We will be removing the whitelist but we need to give people time to clean up their existing codebases with the warning we added. So this will wait for a future major release.

I understand the frustration but would you agree that this slight inconvenience is worth keeping the ecosystem from diverging?

**Comment by Swivelgames** [comment]
@sebmarkbage You're talking about a "slight inconvenience" that's hindering many developers in development, forcing work-arounds and hacks **for nearly three and a half years now**. Suffice to say that the divergence from standards and best practices in order to ensure development can continue on those projects is not entirely an acceptable side-effect to something so imperative. Would you agree?

The few ~40 who have contributed to this thread, among the many many others, will be eagerly waiting for this future major release.

**Comment by gaearon** [comment]
@Swivelgames 

I will refer you to [the section on “stability” in our design principles](https://facebook.github.io/react/contributing/design-principles.html#stability). We know people want this change. This is why we added the deprecation warning for spreading non-DOM props. It generated [a ton of negative feedback](https://twitter.com/search?f=tweets&vertical=default&q=react%20unknown%20prop&src=typd) because people relied on spreading props but we know this is important to people in this thread, and deprecating is a necessary step before we can introduce breakage.

Deprecation only happened in [React 15.2.0](https://github.com/facebook/react/releases/tag/v15.2.0) so it’s only been deployed for half a year. I hope you agree this is a reasonable timeframe to wait for projects to update, including open source projects many people rely on (for example, React Bootstrap and similar libraries).

**Comment by nfroidure** [comment]
Better late then never, i guess. Sadly the issue was raised 3 years ago when a deprecation cycle would not have been necessary if feedback had been taken in count immediately. Not blaming anyone but just giving input on why i decided not to use React in production projects since i noticed no move were done on this issue.

**Comment by gaearon** [comment]
It would indeed be great if all issues could be taken into account immediately. However this is just not how software is developed. At the time, there were always more pressing issues. We try to do our best to prioritize work on what is likely to help most people, and while this particular request has a lot of support, so do many other things we are and have been working on.

I appreciate you sharing your concerns but let’s keep this thread on topic. Yes, it’s a change we wanted to make for a while, and yes, it took us a while to get to add the warning. If you open any popular issue, you will find comments saying that particular issue is the most urgent one.

I’m sorry this particular change prevented you from using React but I hope that you found an alternative that better suits your needs in the meantime. I also hope you’ll reconsider after we make this change.

Cheers!

**Comment by cytGitHub** [comment]
@steida @Swivelgames @janhancic 
`
<video webkit-playsinline=true x-webkit-airplay=true x5-video-player-type = "h5"  playsinline>
`
how to use Video Attribute in React JSX ??

**Comment by OlegLustenko** [comment]
@cytGitHub 
Seems this source is reasonable: https://github.com/zeit/next.js/blob/master/examples/with-amp/pages/_document.js

**Comment by forrest-akin** [comment]
I like the current workaround of using a ref callback to attach your custom attribute. Is there a good way of doing this in a stateless functional component where you can't use refs?

**Comment by funwithtriangles** [comment]
Would it make sense for a recommended approach to be officially documented somewhere? This thread is quite baffling.

**Comment by forrest-akin** [comment]
I've found [this](https://github.com/facebook/react/issues/140#issuecomment-186010590) approach to work best in most cases

**Comment by gaearon** [comment]
I think we’re ready to do it, but I don’t think anybody on the team has time right now. If somebody wants to contribute a big feature to React this is the opportunity 😄 

https://github.com/facebook/react/issues/9477

**Comment by nhunzaker** [comment]
I dug into this a while back:
https://github.com/facebook/react/pull/7311

I'd be happy to see this through, but I'm chewing on a few other outstanding pull requests right now if someone would like to pick up the work. Otherwise I could circle back this once I close out those pull requests

**Comment by gaearon** [comment]
OK, I think rebasing https://github.com/facebook/react/pull/7311 is the most reasonable way forward since you already did a bunch of the work. (Sorry I forgot!) I’ll leave a note in https://github.com/facebook/react/issues/9477 but let’s discuss more specifics there.

**Comment by gaearon** [comment]
After research in #6459, #7311, #10229, #10397, this was fixed by #10385, #10470, https://github.com/facebook/react/pull/10495, https://github.com/facebook/react/pull/10564, and some follow up work to ensure correctness in https://github.com/facebook/react/pull/10536, https://github.com/facebook/react/pull/10543, https://github.com/facebook/react/pull/10559, https://github.com/facebook/react/pull/10588, https://github.com/facebook/react/pull/10594, https://github.com/facebook/react/pull/10595.

Learn more about the new behavior in React 16:
https://facebook.github.io/react/blog/2017/09/08/dom-attributes-in-react-16.html

**Comment by Swivelgames** [comment]
![Thank You](https://media.giphy.com/media/IcGkqdUmYLFGE/giphy.gif)

**Comment by gaearon** [comment]
Thanks to @nhunzaker who did most of the work. :-)

**Comment by everdimension** [comment]
Hmmm, but I still need to allow custom attributes on regular dom elements.

I used to be happy with this:

    import { DOMProperty } from 'react-dom/lib/ReactInjection';

    DOMProperty.injectDOMPropertyConfig({
      isCustomAttribute: attributeName =>
        attributeName.startsWith('cq-') || attributeName.startsWith('stx'),
    });

But in react 16 you no longer expose the `DOMProperty` object although I see you still use it in the source code.

As you write in the [release notes](https://reactjs.org/blog/2017/09/26/react-v16.0.html#packaging):

> There is no react/lib/* and react-dom/lib/* anymore < ... > let us know about your specific case in a new issue, and we’ll try to figure out a migration strategy for you

Well, please do! :)

I rely on a library which, in turn, relies on custom elements (web components). I have to provide layout for it.

While I'm not happy that they chose this path of adding custom attributes to the elements, I feel like providing required markup should be accomplishable.

So in short, is there *still* a way to tell react that certain attributes are okay?

If not, shouldn't this issue be reopened?

**Comment by aweary** [comment]
@everdimension what attributes are you having trouble with? Arbitrary custom attributes should work fine with React 16 by default, which is why this is closed.

**Comment by everdimension** [comment]
Ok, I found out what's going on.

The current behavior is something like this:

    <span stripped-unknown-attribute>attribute will not be rendered</span>
    <span kept-unknown-attribute="">attribute will be rendered</span>

A bit surprising for me, but perhaps intended. In the first case react also gives a warning: 

*Warning: Received `true` for non-boolean attribute `stripped-unknown-attribute`. If this is expected, cast the value to a string.*

If this is intended, I'll update my code accordingly.

So is it?... :)

**Comment by tkrotoff** [comment]
@everdimension same here, I was surprised that custom boolean attributes are not supported.
For examples, [AngularJS Material](https://material.angularjs.org/latest/layout/introduction) and [Angular Flex-Layout](https://github.com/angular/flex-layout/wiki/API-Documentation) :
```HTML
<div layout="row">
  <div flex hide-lg>First item in row</div>
  <div flex show hide-md="{{hideBox}}">Second item in row</div>
</div>
```
There are probably other legit use of custom "boolean" attributes. IMHO they should be supported by React.

**Comment by gaearon** [comment]
They are supported, it just explicitly asks you to pass `"true"` as a string value.

This is because there is no sure way to tell how to apply a boolean attribute. In some cases it's applied by setting an empty value (for true) and removing it (for false). In other cases it expects `"true"` and `"false"` as strings. But again, we can’t always pass them automatically because `"false"` can be interpreted as truthy (because the string is).

This is why you need to explicitly specify them:

```
<div something="" />
<div something="true" />
<div something="false" />
```

depending on how it should be used.

**Comment by Swivelgames** [comment]
@gaearon Would it make sense to simply treat the `boolean` versions as flags but still warn about them? In certain scenarios, just like not passing in custom attributes, supporting libraries and others may rely on the attribute simply existing, not specifically being equal to `"true"`.

**Comment by Swivelgames** [comment]
@gaearon Would it make sense to simply treat the `boolean` versions as flags but still warn about them? In certain scenarios, just like not passing in custom attributes, supporting libraries and others may rely on the attribute simply existing, not specifically being equal to `"true"`.

In those cases, the following would all be considered "true":
```
<div something />
<div something="" />
<div something="true" />
<div something="false" />
```

And the following would be considered "false":
```
<div />
```

I especially think with propTypes, FlowType, TypeScript, and others, attributes used as boolean flags are probably done so very explicitly.

This is also true because it's a bit more cumbersome to check for `"true"`, `"True"`, and `"TruE"` rather than `.hasAttribute()` and other methods.

**Comment by gaearon** [comment]
Then this form works for you, doesn’t it?

```js
<div something="" />
```

I understand the shorthand form feels “nicer”:

```js
<div something />
```

But it might actually be confusing in the future. There are discussions in the JSX spec repo about changing the semantics to compile this to `something={something}`. This would align it with ES6 object shorthand form, and also make it less cumbersome to pass many props down:

```js
const className = 'foo';
const user = props.comment.author;
return <Comment className user />
```

So I wouldn’t recommend relying on the shorthand form anyway. It’s not changing anytime soon, but within a few years—maybe.

**Comment by everdimension** [comment]
Using 

```html
<div something="" />
```

works for me.

According to [html spec](https://www.w3.org/TR/html51/syntax.html#elements-attributes) it is semantically the same as using an empty attribute, the latter being an implicit way to give the value of an empty string.

**Comment by Swivelgames** [comment]
@everdimension @gaearon Indeed! The only difference being that extra conditional logic is necessary in the render methods in order to prevent that attribute from being used in the component because of this.

Again, these are truthy values:
```html
<div something />
<div something="" />
<div something="true" />
<div something="false" />
```

I think the big "issue" here would moreso just be that the following will result in an undesired "truthy" outcome:
```jsx
<MyComponent something={false} />
```

Because instead of rendering it as:
```html
<div />
```

It would be rendered as:
```html
<div something="false" />
```

Creating a sort of "false-positive", when the desired intent would be to prevent the `something` attribute from being attached by setting it to false. In its current state, the true problem here is not with ensuring a truthy value, but instead with intuitively ensuring a false value. `"false"` is truthy, and that's the real issue here.

The component example above is explicit, of course, but in practice it could be `something={randomlySourcedBoolean}`. In the current implementation, the much more verbose and cumbersome solution (especially in the case of large sets of JSX markdown) seems like the only viable one:
```jsx
if (randomlySourcedBoolean) {
  return <MyComponent something />;
} else {
  return <MyComponent />;
}
```

Which is not a very intuitive solution :P

The possibility of a false-positive is definitely high in the case of `{false}` becoming `"false"`, especially since, principally, these values are entirely the opposite of each other. We might as well be setting `{false}` to `"true"`, because through tests these would yield the same result.

We are, after all, specifically talking about boolean attributes, and the current implementation is not consistent with how boolean attributes should work.

>2.5.2 Boolean attributes
>
>A number of attributes are boolean attributes. The presence of a boolean attribute on an element represents the true value, and the absence of the attribute represents the false value.
>
>If the attribute is present, its value must either be the empty string or a value that is an ASCII case-insensitive match for the attribute's canonical name, with no leading or trailing whitespace.
>
>The values "true" and "false" are not allowed on boolean attributes. To represent a false value, the attribute has to be omitted altogether.

If we really want to get into the nitty-gritty, technically shouldn't `something={true}` be passed through as `<div something="something" />`?

I argue on the side of intention specifically. Obviously, setting `something="true"` should absolutely pass through as the same. But boolean attributes are specific types of attributes.

**Comment by Swivelgames** [comment]
@everdimension @gaearon Indeed! That's the principal problem here.

The possibility of a false-positive is definitely high in the case of `{false}` becoming `"false"`, especially since, principally, these values are entirely the opposite of each other. We might as well be setting `{false}` to `"true"`, because through tests these would yield the same result.

We are, after all, specifically talking about boolean attributes, and the current implementation is not consistent with how boolean attributes should work.

[**3.2.2 Boolean attributes**](https://www.w3.org/TR/2008/WD-html5-20080610/semantics.html#boolean)

>A number of attributes in HTML5 are boolean attributes. **_The presence of a boolean attribute on an element represents the true value,_** and **_the absence of the attribute represents the false value._**
>
>If the attribute is present, **_its value must either be the empty string or a value that is a case-insensitive match for the attribute's canonical name, with no leading or trailing whitespace._**

I argue on the side of intention specifically. Obviously, setting `something="false"` should absolutely pass through as the same. But boolean attributes are specific types of attributes; `<div something="" />` is specifically a valid truthy value.

These are truthy values:
```html
<div something />
<div something="" />
<div something="true" />
<div something="false" />
```

I think the big "issue" here would moreso just be that the following will result in an undesired "truthy" outcome:
```jsx
<MyComponent something={false} />
```

Because instead of rendering it as:
```html
<div />
```

It would be rendered as:
```html
<div something="false" />
OR
<div something="" />
```

Creating a sort of "false-positive", when the desired intent would be to prevent the `something` attribute from being attached by setting it to false. In its current state, the true problem here is not with ensuring a truthy value, but instead with intuitively ensuring a false value. `"false"` is truthy, and that's the real issue here.

The component example above is explicit, of course, but in practice it could be `something={randomlySourcedBoolean}`. In the current implementation, the much more verbose and cumbersome solution (especially in the case of large sets of JSX markdown) seems like the only viable one:
```jsx
if (randomlySourcedBoolean) {
  return <MyComponent something />;
} else {
  return <MyComponent />;
}
```

Which is not a very intuitive solution :P

**Comment by aweary** [comment]
@Swivelgames if you do `<div something={false} />` you'll get the warning mentioned earlier about a boolean value with an unknown attribute. It will render it as `<div />`, not `<div something="false" />` ([example](https://jsfiddle.net/cocbr5en/))

To remove a boolean attribute you can use `null` 

```jsx
<div something={null} />
// renders <div />
<div something="" />
// renders <div something />
```

Please refer to the previous issues on this topic for more context:

[Make on/off, yes/no boolean attributes work #10589](https://github.com/facebook/react/issues/10589)
[Decide on the desired behavior when a boolean value is passed to an unknown prop. #10521](https://github.com/facebook/react/issues/10521)
[RFC: Plan for Attributes in React 16](https://github.com/facebook/react/issues/10399)
[Boolean attributes on Web Components #9230](https://github.com/facebook/react/issues/9230)

We understand what the HTML spec says about boolean attributes, and recognize the inconsistency. React is a higher-level abstraction where boolean values can sometimes represent enumerable attributes which have a boolean set of values (`spellcheck` for example), so it's important that we implement a solution with that context in mind.

**Comment by gaearon** [comment]
This thread is getting really long and every message spams a lot of people. The feature has been implemented in React 16.

If you have specific concerns or ideas about it please file a new focused issue. Thanks!
